### PR TITLE
index k-mer coordinates

### PR DESCRIPTION
**Pull request by karasikov**
annotation and compression of k-mer coordinates

### CODE REVIEW

File: metagraph/src/cli/annotate.cpp [hunk 1]
```diff
@@ -324,6 +324,55 @@ void annotate_data(std::shared_ptr<graph::DeBruijnGraph> graph,
...
+                            && config.num_kmers_in_seq + k - 1 != sequence.size()) {
+                        logger->error("All input sequences must have the same"
+                                      " length when flag --const-length is on");
+                        exit(1);
+                    }
+                    if (sequence.size() >= k) {
+                        uint64_t num_kmers = sequence.size() - k + 1;
+                        batcher.push_and_pay(sequence.size(),
+                                             std::move(sequence), std::move(labels), coord);
+                        coord += num_kmers;
```
Comment (hmusta) [inline metagraph/src/cli/annotate.cpp:371]: shouldn't this be `sequence.size()`?

**Review by hmusta — CHANGES_REQUESTED** [review]

File: metagraph/src/cli/config/config.hpp [hunk 2]
```diff
@@ -91,6 +92,7 @@ class Config {
     unsigned int row_diff_stage = 0;
     unsigned int max_path_length = 100;
     unsigned int smoothing_window = 1;  // no smoothing by default
+    unsigned int num_kmers_in_seq = 0;  // assume all input reads have this length
```
[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/cli/config/config.hpp:95]: The current interface of the diff-transform indeed assumes that all columns store paths of the same length.
That's why I'm using this global to enforce that restriction and make it obvious.

File: metagraph/src/annotation/representation/column_compressed/annotate_column_compressed.hpp [hunk 3]
```diff
@@ -63,6 +63,9 @@ class ColumnCompressed : public MultiLabelEncoded<Label> {
     void add_label_counts(const std::vector<Index> &indices,
                           const VLabels &labels,
                           const std::vector<uint64_t> &counts) override;
+    // for each label and index 'i' add numeric attribute 'coord'
```
[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/annotation/representation/column_compressed/annotate_column_compressed.hpp:66]: Hm, it should work that way. Only triplets (k-mer, label, coord) must be unique.

File: metagraph/src/annotation/representation/column_compressed/annotate_column_compressed.cpp [hunk 4]
```diff
@@ -215,18 +233,87 @@ void ColumnCompressed<Label>::serialize(const std::string &filename) const {
...
+
+    uint64_t num_coordinates = 0;
+
+    for (size_t j = 0; j < coords_.size(); ++j) {
+        // sort pairs <rank, coord>
+        auto &c_v = const_cast<ColumnCompressed*>(this)->coords_[j];
+        ips4o::parallel::sort(c_v.begin(), c_v.end(), std::less<>(), get_num_threads());
+        assert(std::unique(c_v.begin(), c_v.end()) == c_v.end());
+        #pragma omp parallel for num_threads(get_num_threads())
+        for (size_t i = 0; i < c_v.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/annotation/representation/column_compressed/annotate_column_compressed.cpp:265]: I guess that should figure that out automatically.

[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/annotation/representation/column_compressed/annotate_column_compressed.cpp:265]: `static` is not required in this case

File: metagraph/src/cli/annotate.cpp (hunk 1 above)
Comment (karasikov) [reply metagraph/src/cli/annotate.cpp:371]: num_kmers works as well but we can add a margin for delimiting

**Pull request merged**